File: src/pathfinder_combat_simulator/core/monster_database.py
import json
import os
from typing import List, Optional, Any, Dict # Added Dict and Any
from .combatant import Combatant

class MonsterDatabase:
    """
    Handles persistent storage and retrieval of monster statblocks
    Implements Part 1.2 of the specification
    """

    def __init__(self, database_path: str = "monster_data"):
        self.database_path = database_path
        # It's good practice to ensure the path is absolute or clearly relative to a known location.
        # For now, assuming it's relative to where the script is run or a predefined base.
        # If this script is part of a larger package, constructing path from package root might be better.
        # Example: os.path.join(os.path.dirname(__file__), '..', database_path) if monster_data is sibling to core
        os.makedirs(self.database_path, exist_ok=True)

    def save_monster(self, combatant: Combatant) -> bool:
        """Save a monster template to JSON file"""
        try:
            # Ensure combatant.name is filesystem-safe.
            # Replace spaces with underscores, convert to lowercase.
            # Consider more robust sanitization if names can have special characters.
            safe_name = combatant.name.lower().replace(' ', '_')
            # Basic sanitization for other potentially problematic characters (very basic example)
            safe_name = "".join(c if c.isalnum() or c in ['_', '-'] else '' for c in safe_name)
            if not safe_name: # Handle cases where name becomes empty after sanitization
                safe_name = "unnamed_monster"

            filename = f"{safe_name}.json"
            filepath = os.path.join(self.database_path, filename)

            with open(filepath, 'w') as f:
                json.dump(combatant.to_dict(), f, indent=2)

            return True
        except Exception as e:
            print(f"Error saving monster {combatant.name}: {e}")
            return False

    def load_monster(self, monster_name: str) -> Optional[Combatant]:
        """Load a monster template from JSON file"""
        try:
            # Handle both "Monster Name" and "monster_name.json" formats
            if monster_name.endswith('.json'):
                filename = monster_name
            else:
                # Apply similar sanitization to monster_name for lookup
                safe_name = monster_name.lower().replace(' ', '_')
                safe_name = "".join(c if c.isalnum() or c in ['_', '-'] else '' for c in safe_name)
                if not safe_name: # Should ideally not happen if names are managed well
                    return None
                filename = f"{safe_name}.json"

            filepath = os.path.join(self.database_path, filename)

            if not os.path.exists(filepath):
                # Try original formatting if sanitized one not found, just in case
                if not monster_name.endswith('.json'):
                    original_filename = f"{monster_name.lower().replace(' ', '_')}.json"
                    original_filepath = os.path.join(self.database_path, original_filename)
                    if os.path.exists(original_filepath):
                        filepath = original_filepath
                    else:
                        print(f"Monster file not found: {filepath} (and other variants like {original_filename})")
                        return None
                else: # if it already ended with .json and wasn't found
                    print(f"Monster file not found: {filepath}")
                    return None

            with open(filepath, 'r') as f:
                data = json.load(f)

            # The Combatant.from_dict method should handle creating the object
            combatant = Combatant.from_dict(data)
            # Reset for combat is good practice if this instance is directly used
            combatant.reset_for_combat()

            return combatant
        except Exception as e:
            print(f"Error loading monster {monster_name}: {e}")
            return None

    def list_monsters(self) -> List[str]:
        """List all available monster templates"""
        try:
            files = [f for f in os.listdir(self.database_path) if f.endswith('.json')]
            # Attempt to revert filenames to more readable names
            # Remove .json, replace underscores with spaces, title case
            # This assumes filenames were created by replacing spaces with underscores.
            monster_names = []
            for f in files:
                name_part = f[:-5] # Remove .json
                # A simple heuristic: if it contains underscores, it was likely space-separated.
                # Otherwise, it might have been a single word or CamelCase.
                # This part is tricky without knowing the exact naming convention used for saving.
                # The original code just did replace('_', ' ').title()
                # Let's stick to that for consistency with original behavior.
                readable_name = name_part.replace('_', ' ').title()
                monster_names.append(readable_name)
            return monster_names
        except Exception as e:
            print(f"Error listing monsters: {e}")
            return []

    def delete_monster(self, monster_name: str) -> bool:
        """Delete a monster template"""
        try:
            # Apply similar sanitization as in load_monster to find the file
            if monster_name.endswith('.json'): # Should not happen if called with readable name
                filename_to_delete = monster_name
            else:
                safe_name = monster_name.lower().replace(' ', '_')
                safe_name = "".join(c if c.isalnum() or c in ['_', '-'] else '' for c in safe_name)
                if not safe_name:
                    print(f"Cannot delete monster with invalid name format: {monster_name}")
                    return False
                filename_to_delete = f"{safe_name}.json"

            filepath = os.path.join(self.database_path, filename_to_delete)

            if os.path.exists(filepath):
                os.remove(filepath)
                return True
            else:
                # Try original formatting before giving up
                original_filename = f"{monster_name.lower().replace(' ', '_')}.json"
                original_filepath = os.path.join(self.database_path, original_filename)
                if os.path.exists(original_filepath):
                    os.remove(original_filepath)
                    return True
                else:
                    print(f"Monster file not found for deletion: {filepath} (and other variants like {original_filename})")
                    return False
        except Exception as e:
            print(f"Error deleting monster {monster_name}: {e}")
            return False

File: src/core/monsterDatabase.ts
import { Combatant, CombatantData } from "./combatant.ts";

function sanitizeNameForFilename(name: string): string {
  if (!name) return "unnamed_monster";
  let safe_name = name.toLowerCase().replace(/\s+/g, '_');
  safe_name = safe_name.replace(/[^a-z0-9_-]/g, ''); // Allow letters, numbers, underscore, hyphen
  return safe_name || "unnamed_monster";
}

export class MonsterDatabase {
  private databasePath: string;

  constructor(databasePath: string = "data") {
    this.databasePath = databasePath;
    // Deno.mkdir is async, ensure directory exists (can be done at first operation or constructor)
    // For simplicity, we'll assume the directory exists or operations will handle it.
    // Or, we can add an async init method if needed.
    // Let's try to ensure it in constructor, making constructor potentially async or using top-level await if main module.
    // For a class constructor, direct async operations are tricky. Better to ensure path exists before operations.
  }

  private async ensureDbPathExists(): Promise<void> {
    try {
      await Deno.mkdir(this.databasePath, { recursive: true });
    } catch (error) {
      if (error instanceof Deno.errors.AlreadyExists) {
        return; // Directory already exists, which is fine
      }
      console.error(`Error creating database directory at ${this.databasePath}:`, error);
      throw error; // Re-throw if it's another error
    }
  }

  async saveMonster(combatant: Combatant): Promise<boolean> {
    await this.ensureDbPathExists();
    const safeName = sanitizeNameForFilename(combatant.name);
    const filename = `${safeName}.json`;
    // Deno.cwd() might not be what we want if the script is run from elsewhere.
    // For Deno, relative paths are usually relative to the current working directory.
    // If this class is imported, `this.databasePath` should be an absolute path or relative to a known root.
    // Assuming `this.databasePath` is correctly resolved (e.g. "data" in project root).
    const filepath = `${this.databasePath}/${filename}`;

    try {
      const data = combatant.toData(); // Use the new method name
      await Deno.writeTextFile(filepath, JSON.stringify(data, null, 2));
      return true;
    } catch (e) {
      console.error(`Error saving monster ${combatant.name} to ${filepath}:`, e);
      return false;
    }
  }

  async loadMonster(monsterName: string): Promise<Combatant | null> {
    // No need to ensureDbPathExists for loading, if it doesn't exist, file won't be found.
    let filenameToTry: string;
    if (monsterName.endsWith('.json')) {
      filenameToTry = monsterName;
    } else {
      const safeName = sanitizeNameForFilename(monsterName);
      filenameToTry = `${safeName}.json`;
    }

    const filepath = `${this.databasePath}/${filenameToTry}`;

    try {
      const fileContent = await Deno.readTextFile(filepath);
      const data = JSON.parse(fileContent) as CombatantData;
      const combatant = Combatant.fromData(data); // Use the new method name
      combatant.resetForCombat(); // Good practice
      return combatant;
    } catch (e) {
      // Try original name if sanitized one failed (Python version had this logic)
      if (!monsterName.endsWith('.json')) {
          const originalSafeName = monsterName.toLowerCase().replace(/\s+/g, '_');
          // A simpler original sanitization for lookup if the first one failed
          const originalFilename = `${originalSafeName}.json`;
          const originalFilepath = `${this.databasePath}/${originalFilename}`;
          if (originalFilename !== filenameToTry) { // Avoid re-trying the same file
            try {
                const fileContent = await Deno.readTextFile(originalFilepath);
                const data = JSON.parse(fileContent) as CombatantData;
                const combatant = Combatant.fromData(data);
                combatant.resetForCombat();
                return combatant;
            } catch (e2) {
                // console.warn(`Monster file not found: ${filepath} (and also not at ${originalFilepath})`, e, e2);
            }
          }
      }
      // If any error (NotFound, JSONParse, etc.)
      if (!(e instanceof Deno.errors.NotFound)) {
        console.error(`Error loading monster ${monsterName} from ${filepath}:`, e);
      }
      return null;
    }
  }

  async listMonsters(): Promise<string[]> {
    await this.ensureDbPathExists(); // Ensure path exists before trying to read from it
    const monsterNames: string[] = [];
    try {
      for await (const dirEntry of Deno.readDir(this.databasePath)) {
        if (dirEntry.isFile && dirEntry.name.endsWith('.json')) {
          const namePart = dirEntry.name.slice(0, -5); // Remove .json
          // Attempt to make it more readable: replace underscores, title case
          const readableName = namePart.replace(/_/g, ' ').replace(/\b\w/g, char => char.toUpperCase());
          monsterNames.push(readableName);
        }
      }
      return monsterNames;
    } catch (e) {
      if (e instanceof Deno.errors.NotFound) {
        // If the directory itself doesn't exist, return empty list.
        return [];
      }
      console.error(`Error listing monsters from ${this.databasePath}:`, e);
      return [];
    }
  }

  async deleteMonster(monsterName: string): Promise<boolean> {
    // No need to ensureDbPathExists for deletion, if it doesn't exist, file won't be found for deletion.
    let filenameToDelete: string;
     if (monsterName.endsWith('.json')) { // Should ideally be called with readable name
      filenameToDelete = monsterName;
    } else {
      const safeName = sanitizeNameForFilename(monsterName);
      filenameToDelete = `${safeName}.json`;
    }
    const filepath = `${this.databasePath}/${filenameToDelete}`;

    try {
      await Deno.remove(filepath);
      return true;
    } catch (e) {
      if (e instanceof Deno.errors.NotFound) {
        // Try original name if sanitized one failed
        if (!monsterName.endsWith('.json')) {
            const originalSafeName = monsterName.toLowerCase().replace(/\s+/g, '_');
            const originalFilename = `${originalSafeName}.json`;
            const originalFilepath = `${this.databasePath}/${originalFilename}`;
            if (originalFilename !== filenameToDelete) {
                 try {
                    await Deno.remove(originalFilepath);
                    return true;
                 } catch (e2) {
                    // console.warn(`Monster file not found for deletion: ${filepath} (and also not at ${originalFilepath})`);
                    return false;
                 }
            }
        }
        // console.warn(`Monster file not found for deletion: ${filepath}`);
        return false;
      }
      console.error(`Error deleting monster ${monsterName} (${filepath}):`, e);
      return false;
    }
  }
}
